File: examples/agent-host/typescript/a2a-adapter.ts
import type { AgentHost } from "./assembly.js"

export type RemoteTaskRequest = {
  runId: string; taskId: string; tenantId: string; owner: string
  idempotencyKey: string; message: string
}

export interface RemoteTaskClient {
  submit(request: RemoteTaskRequest): Promise<Record<string, unknown>>
  getTask(taskId: string, tenantId: string, owner: string): Promise<Record<string, unknown>>
  getArtifact(artifactId: string, tenantId: string, owner: string): Promise<Buffer>
}

export class HostA2AAdapter {
  private readonly runTasks = new Map<string, string>()
  constructor(private readonly host: AgentHost, private readonly client: RemoteTaskClient) {}

  async submit(request: RemoteTaskRequest): Promise<Record<string, unknown>> {
    const bound = this.runTasks.get(request.runId)
    if (bound && bound !== request.taskId) {
      return { success: false, error_code: "checkpoint_mismatch",
        error_message: "一个 run_id 不能绑定多个远程 task", retryable: false }
    }
    this.runTasks.set(request.runId, request.taskId)
    const result = await this.client.submit(request)
    await this.host.events.publish("a2a.task.submitted", {
      runId: request.runId, taskId: request.taskId, tenantId: request.tenantId,
      status: result.status ?? "submitted",
    })
    return result
  }

  async getTask(runId: string, taskId: string, tenantId: string, owner: string): Promise<Record<string, unknown>> {
    if (this.runTasks.get(runId) !== taskId) {
      return { success: false, error_code: "checkpoint_mismatch",
        error_message: "run_id 与远程 task_id 不匹配", retryable: false }
    }
    return this.client.getTask(taskId, tenantId, owner)
  }

  async getArtifact(runId: string, artifactId: string, tenantId: string, owner: string): Promise<Buffer> {
    if (!this.runTasks.has(runId)) throw new Error("run_id 尚未绑定远程 task")
    return this.client.getArtifact(artifactId, tenantId, owner)
  }
}

File: examples/agent-host/python/a2a_adapter.py
"""A2A/Artifact 进入 AgentHost 的最小适配层。

该模块只负责把远程 Task 的身份和事件绑定到 Host 的 run_id；远程传输、认证
和真正的 Artifact 持久化仍由注入的 Client/Store 实现。
"""

from dataclasses import dataclass
from typing import Any, Protocol


@dataclass(frozen=True)
class RemoteTaskRequest:
    run_id: str
    task_id: str
    tenant_id: str
    owner: str
    idempotency_key: str
    message: str


class RemoteTaskClient(Protocol):
    async def submit(self, request: RemoteTaskRequest) -> dict[str, Any]: ...

    async def get_task(self, task_id: str, tenant_id: str, owner: str) -> dict[str, Any]: ...

    async def get_artifact(self, artifact_id: str, tenant_id: str, owner: str) -> bytes: ...


class HostA2AAdapter:
    def __init__(self, host: Any, client: RemoteTaskClient) -> None:
        self.host = host
        self.client = client
        self._run_tasks: dict[str, str] = {}

    async def submit(self, request: RemoteTaskRequest) -> dict[str, Any]:
        bound = self._run_tasks.get(request.run_id)
        if bound and bound != request.task_id:
            return {"success": False, "error_code": "checkpoint_mismatch",
                    "error_message": "一个 run_id 不能绑定多个远程 task", "retryable": False}
        self._run_tasks[request.run_id] = request.task_id
        result = await self.client.submit(request)
        await self.host.events.publish("a2a.task.submitted", {
            "run_id": request.run_id, "task_id": request.task_id,
            "tenant_id": request.tenant_id, "status": result.get("status", "submitted"),
        })
        return result

    async def get_task(self, run_id: str, task_id: str, tenant_id: str, owner: str) -> dict[str, Any]:
        bound = self._run_tasks.get(run_id)
        if bound != task_id:
            return {"success": False, "error_code": "checkpoint_mismatch",
                    "error_message": "run_id 与远程 task_id 不匹配", "retryable": False}
        return await self.client.get_task(task_id, tenant_id, owner)

    async def get_artifact(self, run_id: str, artifact_id: str, tenant_id: str, owner: str) -> bytes:
        if run_id not in self._run_tasks:
            raise ValueError("run_id 尚未绑定远程 task")
        return await self.client.get_artifact(artifact_id, tenant_id, owner)
